sdk/samples/iot/aziot_realtek_amebaD/aziot_realtek_amebaD.ino

```diff
@@ -0,0 +1,379 @@
+static char telemetry_topic[128];
+static uint8_t telemetry_payload[100];
+static uint32_t telemetry_send_count = 0;
+static unsigned char* ca_pem_nullterm;
+
+extern "C"{
+  extern int mbedtls_base64_decode( unsigned char *dst, size_t dlen, size_t *olen, const unsigned char *src, size_t slen);
+  extern int rom_hmac_sha256(const u8 *key, size_t key_len, const u8 *data, size_t data_len, u8 *mac);
+  extern int mbedtls_base64_encode( unsigned char *dst, size_t dlen, size_t *olen, const unsigned char *src, size_t slen);
+}
+static void createNullTerminatedRootCert()
+{
+  ca_pem_nullterm = (unsigned char*)malloc(ca_pem_len + 1);
+
+  if (ca_pem_nullterm == NULL)
+  {
+    Serial.println("Failed allocating memory for null-terminated root ca");
+  }
+  else {}
+  
+  memcpy(ca_pem_nullterm, ca_pem, ca_pem_len);
+  ca_pem_nullterm[ca_pem_len] = '\0';
+}
+
+
+static void connectToWiFi()
+{
+  Serial.begin(115200);
+  Serial.println();
+
+  Serial.print("Connecting to WIFI SSID ");
+  Serial.println(ssid);
+
+  WiFi.begin((char*)ssid, (char*)password);
+  while (WiFi.status() != WL_CONNECTED)
+  {
+    delay(500);
+    Serial.print(".");
+  }
+
+  Serial.println("");
+  Serial.print("WiFi connected, IP address: ");
+  Serial.println(WiFi.localIP());
+}
+
+void receivedCallback(char* topic, byte* payload, unsigned int length)
+{
+  Serial.print("Received [");
+  Serial.print(topic);
+  Serial.print("]: ");
+  for (unsigned int i = 0; i < length; i++)
+  {
+    Serial.print((char)payload[i]);
+  }
+  Serial.println("");
+}
+
+static void initializeClients()
+{
+  Serial.println("Initializing MQTT client");
+  
+  wifi_client.setRootCA(ca_pem_nullterm);
+
+  if (az_result_failed(az_iot_hub_client_init(
+          &hub_client,
+          az_span_create((uint8_t*)host, strlen(host)),
+          az_span_create((uint8_t*)device_id, strlen(device_id)),
+          NULL)))
+  {
+    Serial.println("Failed initializing Azure IoT Hub client");
+    return;
+  }
+
+  mqtt_client.setServer(host, mqtt_port);
+  mqtt_client.setCallback(receivedCallback);
+
+  Serial.println("MQTT client initialized");
+}
+
+int64_t iot_sample_get_epoch_expiration_time_from_minutes(uint32_t minutes)
+{
+
+  long ts = 0;
+  long tus = 0;
+  unsigned int ttk = 0;
+
+  //it should be ok to do init more than one time. It'll handle inside sntp_init().
+  sntp_init();
+
+  sntp_get_lasttime(&ts, &tus, &ttk);
+  
+  while(ts == 0){
+    vTaskDelay(1000 / portTICK_RATE_MS);
+    sntp_get_lasttime(&ts, &tus, &ttk);
+  } 
+
+  return (int64_t)ts + minutes * 60;
+}
+
+static void hmac_sha256_sign_signature(
+    az_span decoded_key,
+    az_span signature,
+    az_span signed_signature,
+    az_span* out_signed_signature)
+{ 
+  if(rom_hmac_sha256(
+    az_span_ptr(decoded_key), 
+    (size_t)az_span_size(decoded_key),
+    az_span_ptr(signature),
+    (size_t)az_span_size(signature),
+    az_span_ptr(signed_signature)) != 0)
+
+  {
+    Serial.println("[ERROR] rom_hmac_sha256 failed");
```

Comment:
return after fail?

Reply:
No, we will keep as is as this was provided by the Realtek team.